File: Octantoid_Reg.m
function [reg,dreg]=calc_regterm_3spha_exp_v2(a,LMAX)
vec=zeros(1,(LMAX+1)^2);
count=1;
for i=0:LMAX
    for j=-i:i
        vec(count)=i;
        count=count+1;
    end
end
vec=vec/LMAX;
xc=a(1:(LMAX+1)^2);
yc=a((LMAX+1)^2+1:2*(LMAX+1)^2);
zc=a(2*(LMAX+1)^2+1:3*(LMAX+1)^2);
reg=(vec.*yc)*yc'+(vec.*zc)*zc';
dreg=[zeros(1,(LMAX+1)^2), 2*vec.*yc 2*vec.*zc];
end

File: octantoid_reg.c
#include"utils.h"


void octantoid_reg(double *a,int LMAX,double *oreg,double *doreg)
{
    /*
     * Octantoid regularization
     * INPUT:
     * a octantoid representation
     * oreg pointer to double
     * doreg 3*(LMAX+1)^2 array
     */
    double reg=0;
    int al=pow(LMAX+1,2);
    int count=0;
    double *vec=calloc(al,sizeof(double));
    for(int i=0;i<=LMAX;i++)
        for(int j=-i;j<=i;j++)
        {
            vec[count]=(double)i/LMAX;
            count++;
        }
    double *xc=a;
    double *yc=a+al;
    double *zc=a+2*al;
    for(int i=0;i<al;i++)
    {
        reg+=vec[i]*yc[i]*yc[i]+vec[i]*zc[i]*zc[i];
        doreg[al+i]=2*vec[i]*yc[i];
        doreg[2*al+i]=2*vec[i]*zc[i];
    }
    *oreg=reg;
    free(vec);
}
